### DISCUSSION TITLE
Extending TFT_eSPI library

### DISCUSSION THREAD

**Discussion by FrankBoesing** [Bodmer/TFT_eSPI#1536] [New processor/display support or capability request]
Hi,

i'd like to extend your great library. Some needed functions are "private", so this is not possible. Changing that "private:" into "protected:" would allow to use inheritance and to add extensions on top of your lib.

Is there a chance that you change this?

Thank you in advance,   
Frank

**Comment by FrankBoesing** [comment]
p.s. and could you add non-inlined versions of begin\_tft\_write() / end\_tft\_write(), please?

**Comment by Bodmer** [comment]
Maybe, a few provisos:
1. The extensions must be published online for Arduino users - explain what is proposed.
2. I suspect not all need to be removed from private - which ones?
3. Code is inline for performance so there would need to be a default inline, not sure how this can be changed in a tidy way without another #define.

**Comment by FrankBoesing** [comment]
Hi, yes it is intended to publish it /them
A first extension is to add the TTF fonts used by the ILI9341_t3 library ((c) Paul Stoffregen and others). Years ago I converted >900 Google fonts to it: https://github.com/FrankBoesing/fonts/tree/master/ofl - they display pretty fast.

I have uploaded an example here: https://github.com/FrankBoesing/WT32-SC01-TTF-TEST - it includes your library, slightly edited.
At the moment, it only needs begin_tft_write() / end_tft_write() - i've added non-inlined versions.

**Comment by Bodmer** [comment]
Hi, thanks, that is great.

I have run your example with a few minor tweaks on an RP2040 using the stock library. It runs very well.

I used [startWrite() and endWrite()](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.cpp#L3454-L3475) as these are the user interface for private begin_tft_write and end_tft_write. I need to think on this further as there may be a DMA transaction iin progress when the TTF font call is made and this needs to wait. At the moment I think a dedicated start (with DMA wait) and end functions may be needed.

**Comment by FrankBoesing** [comment]
No hurry, take your time :)
I'll use startWrite() until then.

**Comment by FrankBoesing** [comment]
I'd like to add some methods to make a nice tool work - I modified it to work with other screen-sizes (than ILI9341). It's a PowerPoint :) with a macro that converts the screens to tft display commands.... you can create your GUI with powerpoint, start the macro and get a (almost) 1:1 sourcecode that displays this gui on the tft...

This is the original: https://github.com/KrisKasprzak/Powerpoint-ILI9341_t3

**Reply by Bodmer** [reply]
Interesting. Thanks for the link. I patched together Kris's sketch and your example and it works fine apart from the button text since that uses a different font render call. This PowerPoint idea is rather neat, however I could not run the macro, possibly because I have an older version of PowerPoint (2010 !).

**Reply by Bodmer** [reply]
I have added the option in v2.4.24 to optionally use the "print" method in the button class, so now your extension example correctly plots the font in the button.

Just a simple change is needed to your example code to invoke the behaviour:
```
  void setTTFFont(const tftfont_t &f)
  {
    font = &f;
    setTextFont(255);
  }
```

**Comment by FrankBoesing** [comment]
Thanks :)
Today, I spent some time to work on the Powerpoint makro. It is a pity that it does not work for you!
I'll create a repository, tomorrow, with a new version that support all displays, auto-adjusts the slide-size, better font handling, and optimizes the output (i.e. dont set the color & font if unchanged, and reduce duplicate comments).
I'll upload a first version of my "font extension" lib for your tft_eSPI, too.

**Comment by FrankBoesing** [comment]
Hm.. however... i just took a look... it seems my version is Powerpoint 2007 (?)

Can you post the errormessages?

**Reply by Bodmer** [reply]
It just reports that the file is corrupted and cannot be loaded. Maybe I'll download and try again.

**Reply by Bodmer** [reply]
OK, security blocked the pptm file type - just had to disable the security lock on it - running now.

**Comment by FrankBoesing** [comment]
Great news :) I'll publish my version tomorrow..

**Reply by Bodmer** [reply]
OK, great.

I have added [filled H/V gradient rectangles](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.h#L460-L463) for compatibility.

**Reply by Bodmer** [reply]
Here is my test sketch that runs under the Arduino environment:
[PowerPointGUI_2_VB_Test.zip](https://github.com/Bodmer/TFT_eSPI/files/7833945/PowerPointGUI_2_VB_Test.zip)


One thing to fix is to get rid of this warning:
```
...PowerPointGUI_2_VB_Test.ino:40:76: warning: ISO C++ forbids converting a string constant to 'char*'...
```
An approach is needed to work out a simple edit the sketch to allow user defined values to be updated on the screen.

**Reply by FrankBoesing** [reply]
I had added the gradients too... you're too fast for me :) OK, I remove my version from the lib.

> An approach is needed to work out a simple edit the sketch to allow user defined values to be updated on the screen.

Yes! But at the moment, I have no good Idea..

**Comment by Bodmer** [comment]
I had a go at tweaking the macro code myself.  I expect you noticed there were a few bugs that needed fixing. TFT_eSPI handles ellipses so there is no need to assume a circle is drawn and I have updated the code. I have created a test slide but it is not complete yet. [Here is  a copy](https://github.com/Bodmer/PowerPoint_to_sketch) FWIW.

**Comment by FrankBoesing** [comment]
Hi, i've uploaded my version.
It generates a .h file now - take a look at the makro. So far, it's not tested much. I found some issues, but I the code is so different that it would be better if you apply the bugfixes yourself and make a pullrequest with your changes :)

https://github.com/FrankBoesing/PowerPointToTFT

**Comment by FrankBoesing** [comment]
A first version of the library is here:
https://github.com/FrankBoesing/TFT_eSPI_ext

**Comment by FrankBoesing** [comment]
Hm,, looks like we have to add a scaling to the fonts-sizes, too. They are too small with a 480x320 display.

**Reply by Bodmer** [reply]
OK. I will have a look at your library later today. At the moment I can see it will not be recognised as an Arduino library as their are some rules to follow. The fonts you have created look good and characters are well formed, I will have to try the Perl script later too.

**Comment by Bodmer** [comment]
I've been experimenting with a way to signal variables in tables and to generate the appropriate lines of code. Example:
![image](https://user-images.githubusercontent.com/15824805/148683242-3e8071c7-93a6-4a1e-9810-cc8a034f6a4e.png)

This generates code like this that will need to be editted by the user:
```
Display.fillRect(65, 103, 61, 29, 65405);
Display.setTextColor(0);
Display.setTTFFont(Arial_8);
Display.setCursor(75, 114 );
static int tab11 = 0;
Display.print(tab11);

Display.fillRect(126, 103, 65, 29, 65405);
Display.setTextColor(0);
Display.setTTFFont(Arial_8);
Display.setCursor(136, 114 );
static float tab21 = 0.00;
Display.print(tab21);

Display.fillRect(191, 103, 68, 29, 65405);
Display.setTextColor(0);
Display.setTTFFont(Arial_8);
Display.setCursor(201, 114 );
const char tab31[] = "Text";
Display.print(tab31);
'''

**Comment by FrankBoesing** [comment]
I think the font sizes on the slast sliede of kris fit only by chance for the 320*240 ili9341.
I think we may need to convert the sizes to "dot". What do you think, does it make sense?

With 480x320 the fontsize 8 in the "screen2" table is ok, but the headline is way too small and should roughly double the size...

**Comment by Bodmer** [comment]
Yes, the screen w and h is 1.5x so maybe that is a good scale factor for a 320x480 screen.

**Comment by FrankBoesing** [comment]
My version calcs the scale factor with the help of screen size and display size to get the "dot"s needed by Powerpoint for all elements und uses the font size the slides uses. However, this using that fontsize seems not to be correct, although PP scales the fonts.... the x1.5 seems not to be correct, too, as it is too less for the headline and too large for the 8pt fonts in the table. Have to think about it...

**Reply by Bodmer** [reply]
The answer may be to change the page size in PowerPoint as it seems to be custom:

![image](https://user-images.githubusercontent.com/15824805/148695475-4dbc5b85-ec68-437f-9193-70be76af0ef4.png)

**Reply by Bodmer** [reply]
I increased the page dimensions manually and the fonts are then automatically scaled for existing pages, but end up at odd values so have to be manually tweaked.

Maybe there is a way to set the page size from the macro?

**Reply by Bodmer** [reply]
Ah, your macro does change the page size, when I manually change the page size the fonts are scaled. When I run the macro thenit changes the page size but the fonts are not scled down so the page then looks bad on the PC screen...

Tests suggest that the point size produced in the sketch is only 1 point down on a good match for the smaller (<20 size) characters.

**Comment by FrankBoesing** [comment]
No, headline is 13 but should be ~25 - fontsize in the table is ok.
I think I know the reason. Font sizes are in px - but windows fontsize is "dot".
Will try to fix that tomorrow evening...

**Comment by FrankBoesing** [comment]
Still doesn't look "perfect", but indeed the 1.5 scalefactor seems almost ok.. have updated the ppt.

**Reply by Bodmer** [reply]
I am getting a very close match with the following approach.

The problem is that there is a scale factor between the pixel and the font point size so that factor needs to be determined, This is a constant for all screen sizes.  I found a scale factor of SF = 1.385 is very close.  The page size is then set to say 479/SF and 319/SF. All pixel based coords are then scaled up by SF to bring them back in range (e.g. 0-479 for width). The fonts are then rendered the correct relative size to a graphics box. I also found I had to increase the cursor y position by fontSize/3 to get the text in the box. Now all other slides look good too and strings are very close to being the correct length. Because the fonts are digitised it can never be perfect due to pixel level rounding, especially on small fonts.

Screen (captured with a cheap webcam!)
![image](https://user-images.githubusercontent.com/15824805/148709477-5540a05b-aaa5-4222-a32d-e89aed5923c9.png)
Master slide:
![image](https://user-images.githubusercontent.com/15824805/148709530-7308c61f-4cb2-4510-b79d-98ed56bce9a2.png)

**Reply by Bodmer** [reply]
Note those green rectangles are sepatate graphics objects to the text (unfilled rectangles) so their size is in pixels rather than font point size.

**Comment by Bodmer** [comment]
My latest version 1.0.4 of the sketch generator is here, it includes test slides only at the moment:
https://github.com/Bodmer/PowerPoint_to_sketch

This includes the first attempt at creating boxes with associated variables.

**Reply by FrankBoesing** [reply]
> My latest version 1.0.4 of the sketch generator is here, it includes test slides only at the moment: https://github.com/Bodmer/PowerPoint_to_sketch
> 
> This includes the first attempt at creating boxes with associated variables.

The link does not work for me :)

**Reply by Bodmer** [reply]
Whoops! Made it public, so link should work now.

**Comment by FrankBoesing** [comment]
Looks good.
It doesn't make much sense for us to work on different versions. It's also difficult to apply your changes each time, and I honestly don't have much desire for this manual effort. I have changed too much.

I'll wait for your version and do other things.
Wish you good luck, and have fun! I look in some time again and take over your version then. Thank you!

Frank

**Reply by Bodmer** [reply]
Agreed, it does not make much sense to have diverging versions.  My version is really just to test ideas and see what works since I have not tried this before or used VB.

I think step 1 should be to focus on the font library extension as that is key to getting the PowerPoint approach working. I will clone your copy of the TFT_ext library and create a list of comments for consideration. OK?

**Reply by FrankBoesing** [reply]
> Agreed, it does not make much sense to have diverging versions. My version is really just to test ideas and see what works since I have not tried this before or used VB.
> 
> I think step 1 should be to focus on the font library extension as that is key to getting the PowerPoint approach working. I will clone your copy of the TFT_ext library and create a list of comments for consideration. OK?

Sure :)

Note, if you use 
```
tx1 = Round((Shp.Left + Shp.Table.Rows(i).Cells(j).Shape.TextFrame.TextRange.BoundLeft) * SF)                  
ty1 = Round((Shp.Top + Shp.Table.Rows(i).Cells(j).Shape.TextFrame.TextRange.BoundTop + 3) * SF) ' Todo: where comes the 3 from?
```
you'll get centered texts in tables for free :) Perhahps right-aligned and for other texts, too - but I have not tested this.

**Reply by Bodmer** [reply]
Thanks for the tip!

**Comment by FrankBoesing** [comment]
I've compiled ttf2bdf for windows:
https://github.com/FrankBoesing/TFT_eSPI_ext/tree/main/extras/ttf2bdf_windows

**Reply by Bodmer** [reply]
That's useful, thanks.

**Comment by Bodmer** [comment]
I have updated the VBA macro and put the updated pptm here:
https://github.com/Bodmer/Powerpoint-ILI9341_t3

I have created a pull request to Kris's repository.

**Comment by FrankBoesing** [comment]
I'll take a look :)

**Comment by Bodmer** [comment]
I have added a ReadMe to the repository here:

https://github.com/Bodmer/PowerPoint_to_sketch
